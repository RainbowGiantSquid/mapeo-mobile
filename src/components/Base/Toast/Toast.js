// @flow
import React from 'react';
import { Modal } from 'react-native';

type Props = {
  children: any,
  onRequestClose?: Function,
  onHide?: Function,
  visible?: boolean
};

type State = {
  visible: boolean
};

class Toast extends React.PureComponent<Props, State> {
  state = { visible: true };
  timer: any;

  componentDidMount() {
    const { onHide } = this.props;
    this.timer = setTimeout(() => {
      this.setState({ visible: false }, () => {
        if (onHide) {
          onHide();
        }
      });
    }, 2000);
  }

  componentDidUpdate(prevProps: Props, prevState: State) {
    if (this.props.visible !== prevState.visible) {
      this.setState({ visible: this.props.visible });
    }
  }

  componentWillUnmount() {
    if (this.timer) {
      clearTimeout(this.timer);
    }
  }

  render() {
    const { children, onRequestClose } = this.props;
    const { visible } = this.state;

    return (
      <Modal
        animation="slide"
        transparent
        visible={visible}
        onRequestClose={onRequestClose || (() => {})}
      >
        {children}
      </Modal>
    );
  }
}

export default Toast;
