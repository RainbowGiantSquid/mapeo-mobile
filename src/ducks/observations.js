// @flow
import { create } from '@src/lib/redux';
import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { createSelector } from '@src/lib/selector';
import { StoreState } from '@types/redux';
import { Observation } from '@types/observation';

export const {
  type: OBSERVATION_LIST,
  action: observationList,
  reducer: observationListReducer,
} = create('OBSERVATION_LIST', {
  success: (state, action) => {
    const newState = update(state, {
      observations: { $set: keyBy(action.payload, 'id') },
    });

    return newState;
  },
});

export const selectContact = createSelector(
  [
    (state: StoreState, id: string): Observation => state.app.observations[id],
  ],
  (observation: Observation): Observation => observation,
  (observation: Observation, id: string): string => id,
);

export default [observationListReducer];
