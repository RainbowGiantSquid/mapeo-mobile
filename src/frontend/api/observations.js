import fixture from "./test_observations.json";

export default class ObservationsApi {
  constructor(baseUrl) {
    this.baseUrl = baseUrl;
  }
  list(cb) {
    setTimeout(() => cb(null, fixture), 2000);
  }
  update() {}
  create() {}
}
