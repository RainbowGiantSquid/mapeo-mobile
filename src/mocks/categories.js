// @flow

export const createCategory = (category?: any) => ({
  id: 'categoryID',
  name: 'categoryName',
  fieldIds: [7],

  ...category
});
