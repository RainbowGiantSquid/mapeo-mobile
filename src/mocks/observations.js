// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 'ObservationID',
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: 'Thu Apr 05 2018 23:06:45 GMT+0200 (CEST)',
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: null,

  ...observation
});
