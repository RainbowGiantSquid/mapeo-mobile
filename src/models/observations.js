// @flow
import type {
  Observation,
  ServerObservationCreate,
  ServerObservationResponse
} from '../types/observation';

export const createDefaultObservation = (): Observation => ({
  created_at: new Date().toISOString(),
  notes: '',
  categoryId: '',
  attachments: [],
  fields: []
});

// Properties at the top level of a server observation object
const SERVER_PROPS: Array<$Keys<ServerObservationResponse>> = [
  'id',
  'version',
  'created_at',
  'timestamp',
  'type',
  'lon',
  'lat',
  'schemaVersion',
  'ref',
  'metadata',
  'attachments',
  'fields',
  'tags'
];

// Properties at top level of local observation object copied from server obj
const LOCAL_PROPS: Array<$Keys<Observation>> = [
  'id',
  'version',
  'created_at',
  'lat',
  'lon',
  'fields'
];

export const parseObservationRequest = (
  observation: Observation
): ServerObservationCreate | ServerObservationResponse => {
  const serverObservation: Object = { tags: {} };

  // Maintain top-level props and copy other props as tags
  Object.keys(observation).forEach(prop => {
    if (SERVER_PROPS.includes(prop)) {
      serverObservation[prop] = observation[prop];
    } else {
      serverObservation.tags[prop] = observation[prop];
    }
  });

  // set field answers as individual tags
  observation.fields.forEach(field => {
    serverObservation.tags[field.name] = field.answer;
  });

  serverObservation.attachments = (observation.attachments || []).map(a => ({
    id: a
  }));

  serverObservation.type = 'observation';

  return serverObservation;
};

export const parseObservationResponse = (
  serverObservation: ServerObservationResponse
): Observation => {
  const obs = createDefaultObservation();

  Object.keys(serverObservation).forEach(prop => {
    if (LOCAL_PROPS.includes(prop)) {
      // $FlowFixMe
      obs[prop] = serverObservation[prop];
    }
  });

  obs.attachments = (serverObservation.attachments || []).map(a => a.id);

  const tags = serverObservation.tags || {};
  obs.notes = tags.notes || '';
  obs.categoryId = tags.categoryId || '';

  return obs;
};
