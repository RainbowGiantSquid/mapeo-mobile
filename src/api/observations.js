// @flow
import { jsonRequest } from './base';
import { keyBy } from 'lodash';
import { parseObservationResponse } from '../models/observations';
import querystring from 'querystring'

import type {
  Observation as ObservationType,
  ServerObservationResponse,
  ServerObservationCreate
} from '../types/observation';

class Observation {
  static list = (opts) =>
    jsonRequest({
      method: 'GET',
      route: '/observations?' + querystring.stringify(opts)
    }).map(observations => {
      var payload = observations.map(parseObservationResponse)
      return keyBy(payload, 'id')
    });

  static create = (observation: ServerObservationCreate) =>
    jsonRequest({
      method: 'POST',
      route: '/observations',
      body: observation
    }).map(parseObservationResponse);

  static update = (observation: ServerObservationResponse) =>
    jsonRequest({
      method: 'PUT',
      route: `/observations/${observation.id}`,
      body: observation
    }).map(parseObservationResponse);
}

export default Observation;
