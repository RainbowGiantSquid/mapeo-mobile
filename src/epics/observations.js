// @flow
import React from 'react';
import { Image } from 'react-native';
import type { ActionsObservable } from 'redux-observable';
import { OBSERVATION_LIST, observationList } from '../ducks/observations';
import { Action } from '../types/redux';
import { Observation } from '../types/observation';

export const observationListEpic = (
  action$: ActionsObservable<Action<string, Observation[]>>,
  store: any
) =>
  action$
    .ofType(OBSERVATION_LIST)
    .filter(
      action =>
        action.status === 'Start' &&
        store.getState().app &&
        store.getState().app.observations &&
        !store.getState().app.observations.length
    )
    .map(() =>
      observationList('', [
        {
          type: 'Old houses and farms',
          id: 1,
          lat: 0.106831547,
          lon: -77.42499473,
          link: 'link',
          created: new Date('2018-03-25T12:00:00Z'),
          name: 'Old houses and farms',
          notes: `Pedro and Maria's house`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/1-house.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Hunting sites',
          id: 2,
          lat: 0.088170932,
          lon: -77.44020885,
          link: 'link',
          created: new Date('2018-03-25T12:30:00Z'),
          name: 'Hunting sites',
          notes: `Don Pancho's hunting camp`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/4-camp.jpg')
            }
          ],
          icon: require('../images/categories/category_2.png'),
          mock: true
        },
        {
          type: 'Threat',
          id: 3,
          lat: 0.112411067,
          lon: -77.41378756,
          link: 'link',
          created: new Date('2018-02-03T17:30:00Z'),
          name: 'Threat',
          notes: `Oil spill 3 February 2018. Ruptured pipeline. Dead fish, oily water`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/5-threat.jpg')
            }
          ],
          icon: null,
          mock: true
        },
        {
          type: 'Medicinal plants',
          id: 4,
          lat: 0.09749025,
          lon: -77.43614404,
          link: 'link',
          created: new Date('2018-02-04T17:30:00Z'),
          name: 'Medicinal plants',
          notes: `Salola fruit plant - used in tea to heal inflamation`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/6-plant.jpg')
            }
          ],
          icon: require('../images/categories/category_6.png'),
          mock: true
        },
        {
          type: 'Waterfalls',
          id: 5,
          lat: 0.105818148,
          lon: -77.4408037,
          link: 'link',
          created: new Date('2018-03-04T17:30:00Z'),
          name: 'Waterfalls',
          notes: `Sacred waterfall, visited during vision quests for ancestors, for taking ayahuasca`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/7-waterfall.jpg')
            }
          ],
          icon: require('../images/categories/category_13.png'),
          mock: true
        },
        {
          type: 'Rivers and streams',
          id: 6,
          lat: 0.119004333,
          lon: -77.42374225,
          link: 'link',
          created: new Date('2018-03-24T17:30:00Z'),
          name: 'Rivers and streams',
          notes: `Point where pipeline crosses river`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/8-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_0.png'),
          mock: true
        },
        {
          type: 'Historical sites',
          id: 7,
          lat: 0.082856811,
          lon: -77.4429836,
          link: 'link',
          created: new Date('2018-03-24T12:30:00Z'),
          name: 'Historical sites',
          notes: `There used to be a village here in the 1940s, it was abandoned after an epidemic of flu which killed most of the elders and children.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/9-specialsite.jpg')
            }
          ],
          icon: require('../images/categories/category_17.png'),
          mock: true
        },
        {
          type: 'Gold',
          id: 8,
          lat: 0.108605553,
          lon: -77.41560494,
          link: 'link',
          created: new Date('2018-03-11T17:30:00Z'),
          name: 'Gold',
          notes: `Illegal gold miners 11 March 2018, excavators, dredges, open tailing pond, mercury.`,
          observedBy: 'Aliya',
          media: [
            {
              type: 'LocalPhoto',
              source: require('../images/mock-data/10-threat.jpg')
            }
          ],
          icon: require('../images/categories/category_8.png'),
          mock: true
        }
      ])
    );

export default [observationListEpic];
