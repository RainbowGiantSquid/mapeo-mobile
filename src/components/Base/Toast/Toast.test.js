// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import Toast from './Toast';
import { createObservation } from '../../../mocks/observations';

describe('Toast tests', () => {
  test('snapshots', () => {
    const children = null;
    const onRequestClose = jest.fn();
    const onHide = jest.fn();

    const props = [{ children }];

    let tree;
    props.forEach(p => {
      tree = renderer
        .create(
          <Toast
            children={p.children}
            onRequestClose={onRequestClose}
            onHide={onHide}
          />
        )
        .toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
