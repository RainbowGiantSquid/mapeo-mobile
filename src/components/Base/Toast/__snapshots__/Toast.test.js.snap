// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast tests snapshots 1`] = `
<Modal
  animation="slide"
  hardwareAccelerated={false}
  onRequestClose={[Function]}
  transparent={true}
  visible={true}
/>
`;
